exp/controllers/gcpmachinepool_controller.go

```diff
@@ -0,0 +1,290 @@
+	"sigs.k8s.io/cluster-api/util/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// GCPMachinePoolReconciler reconciles a GCPMachinePool object and the corresponding MachinePool object.
+type GCPMachinePoolReconciler struct {
+	client.Client
+	ReconcileTimeout time.Duration
+	WatchFilterValue string
+}
+
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=gcpmachinepools,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=gcpmachinepools/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=gcpmachinepools/finalizers,verbs=update
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=gcpmachinepoolmachines,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=gcpmachinepoolmachines/status,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=gcpmachinepoolmachines/finalizers,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=bootstrap.cluster.x-k8s.io,resources=kubeadmconfigs;kubeadmconfigs/status,verbs=get;list;watch
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinepools;machinepools/status,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups="",resources=events,verbs=get;list;watch;create;update;patch
+// +kubebuilder:rbac:groups="",resources=secrets;,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *GCPMachinePoolReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	log := log.FromContext(ctx).WithValues("controller", "GCPMachinePool")
+
+	gvk, err := apiutil.GVKForObject(new(infrav1exp.GCPMachinePool), mgr.GetScheme())
+	if err != nil {
+		return errors.Wrapf(err, "failed to find GVK for GCPMachinePool")
+	}
+
+	c, err := ctrl.NewControllerManagedBy(mgr).
+		WithOptions(options).
+		For(&infrav1exp.GCPMachinePool{}).
+		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(log, r.WatchFilterValue)).
+		Watches(
+			&expclusterv1.MachinePool{},
+			handler.EnqueueRequestsFromMapFunc(machinePoolToInfrastructureMapFunc(gvk)),
+		).
+		// watch for changes in KubeadmConfig to sync bootstrap token
+		Watches(
+			&kubeadmv1.KubeadmConfig{},
```

Comment:
Would we be able to avoid having to watch `KubeadmConfig`? This would prevent anyone enabling `MachinePool` from selecting the bootstrap provider of their choice and force them to use Kubeadm. This is something we've realized happens in CAPZ and there's an open issue to address [this](https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/4854). Other providers, like [CAPA](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/exp/controllers/awsmachinepool_controller.go), don't have this dependency on Kubeadm.

Reply:
A great point @salasberryfin :+1:

Reply:
+1, thanks for catching this!

Reply:
This is a great point, I'd love to abstract the mechanism away from the kubeadm bootstrap provider lock-in. Any idea how we'd enqueue a reconcile from a token rotation when two different bootstrap providers are being utilized 🤔